kratos/processes/apply_periodic_boundary_condition_process.cpp

```diff
@@ -0,0 +1,357 @@
+                    mrMasterModelPart.CreateNewMasterSlaveConstraint("LinearMasterSlaveConstraint", current_num_constraint++, master_node, r_var_y, rSalveNode, r_var_y, master_weight * mTransformationMatrixVariable(1,1), constant_y);
+                    mrMasterModelPart.CreateNewMasterSlaveConstraint("LinearMasterSlaveConstraint", current_num_constraint++, master_node, r_var_z, rSalveNode, r_var_y, master_weight * mTransformationMatrixVariable(1,2), constant_y);
+
+                    if (TDim == 3)
+                    {
+                        mrMasterModelPart.CreateNewMasterSlaveConstraint("LinearMasterSlaveConstraint", current_num_constraint++, master_node, r_var_x, rSalveNode, r_var_z, master_weight * mTransformationMatrixVariable(2,0), constant_z);
+                        mrMasterModelPart.CreateNewMasterSlaveConstraint("LinearMasterSlaveConstraint", current_num_constraint++, master_node, r_var_y, rSalveNode, r_var_z, master_weight * mTransformationMatrixVariable(2,1), constant_z);
+                        mrMasterModelPart.CreateNewMasterSlaveConstraint("LinearMasterSlaveConstraint", current_num_constraint++, master_node, r_var_z, rSalveNode, r_var_z, master_weight * mTransformationMatrixVariable(2,2), constant_z);
+                    }
+                }
+
+            master_index++;
+        }
+    }
+
+    template <int TDim>
+    void ApplyPeriodicConditionProcess::ConstraintSlaveNodeWithConditionForScalarVariable(NodeType& rSalveNode, const GeometryType& rHostedGeometry, const VectorType& rWeights, const std::string& rVarName )
+    {
+        const VariableType r_var = KratosComponents<VariableType>::Get(rVarName);
+
+        IndexType master_index = 0;
+        for (auto& master_node : rHostedGeometry)
+        {
+            int current_num_constraint = mrMasterModelPart.NumberOfMasterSlaveConstraints();
+
+            const double master_weight = rWeights(master_index);
+            #pragma omp critical
+            {
+                mrMasterModelPart.CreateNewMasterSlaveConstraint("LinearMasterSlaveConstraint", current_num_constraint++, master_node, r_var, rSalveNode, r_var, master_weight, 0.0);
+            }
+            master_index++;
+        }
+    }
+
+    void ApplyPeriodicConditionProcess::CalculateTransformationMatrix()
+    {
+        if (mType == "translation"){
+            CalculateTranslationMatrix(-1*mModulus, mTransformationMatrix);
+            CalculateTranslationMatrix(0.0, mTransformationMatrixVariable);
+        }
+        else if (mType == "rotation"){
+            CalculateRotationMatrix(-1*mTheta, mTransformationMatrix);
+            CalculateRotationMatrix(mTheta, mTransformationMatrixVariable);
+        }
+    }
+
+    void ApplyPeriodicConditionProcess::CalculateTranslationMatrix(const double Modulus, MatrixType& rMatrix)
```

Comment:
I am a little lost. Why a 3*3 matrix with Identity in the first 3 rows and cols?

Reply:
For a matrix to induce pure translation in a given direction, the transformation matrix looks like this. 
Step 1 in the following explanation. 
http://paulbourke.net/geometry/rotate/

Reply:
I think this is the same as how opengl handles transformations: https://www.glprogramming.com/red/chapter03.html#name2

Reply:
Hi guys, first of all i wrote it wrongly. I don't understand why it is 4*4, 3*3 would look ok to me.

I mean if Rx Ry Rz are 3*3 matrices rotating around x y and z

than you can obtain a viable rotation as R = Rz*Ry*Rx (of course the result will depend on the order of applicaiton of the rotations)

Reply:
ultimately i guess that this is an internal thing, so as long as the user doesn't see it i can sleep with the 4*4 matrix. 

however...the T matrix is ultimately 3*3 correct? it connects 3 displacements (or velocities) with other 3. If that is so, than the 4*4 is just an intermediate tool to get T. Is that so?

Reply:
yes riccardo.

Reply:
ok then, i resolve this thread